### PR TITLE
Remove Stored Payment Method

### PR DESCRIPTION
**Pull request by ozgur00**

### CODE REVIEW

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt [hunk 1]
```diff
@@ -28,20 +32,36 @@ import com.adyen.checkout.dropin.ui.paymentmethods.PaymentMethodListItem.Compani
 
 @SuppressWarnings("TooManyFunctions")
 class PaymentMethodAdapter(
-    private val paymentMethods: List<PaymentMethodListItem>,
-    private val imageLoader: ImageLoader
+    private val paymentMethods: MutableList<PaymentMethodListItem>,
+    private val imageLoader: ImageLoader,
+    private val onUnderlayExpandListener: ((AdyenSwipeToRevealLayout) -> Unit)? = null
```
Comment (caiofaustino) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt:37]: Is this `onUnderlayExpandListener` ever being used? Why do we need to know if the underlay was expanded in this use case?

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodListDialogFragment.kt [hunk 2]
```diff
@@ -162,4 +183,12 @@ class PaymentMethodListDialogFragment : DropInBottomSheetDialogFragment(), Payme
         val paymentComponentState = GenericComponentState(paymentComponentData, true, true)
         protocol.requestPaymentsCall(paymentComponentState)
     }
+
+    private fun collapseNotUsedUnderlayButtons(recyclerView: RecyclerView, draggedItem: AdyenSwipeToRevealLayout) {
```
Comment (caiofaustino) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodListDialogFragment.kt:187]: Can you explain why this is needed? I think I'm missing something.

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/stored/PreselectedStoredPaymentMethodFragment.kt [hunk 3]
```diff
@@ -137,23 +148,43 @@ class PreselectedStoredPaymentMethodFragment : DropInBottomSheetDialogFragment()
...
+                    binding.storedPaymentMethodItem.textViewText.text = it.name
+                    binding.storedPaymentMethodItem.textViewDetail.visibility = View.GONE
+                    imageLoader.load(it.imageId, binding.storedPaymentMethodItem.imageViewLogo)
                 }
             }
-        )
+        }
+    }
+
+    private fun showRemoveStoredPaymentDialog() {
```
Comment (caiofaustino) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ui/stored/PreselectedStoredPaymentMethodFragment.kt:172]: I thought it was not gonna be possible to delete the PM in this screen. Did you check this with Arjen? Seems very counter intuitive.

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodListDialogFragment.kt (hunk 2 above)
Comment (ozgur00) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodListDialogFragment.kt:187]: Whenever shopper swipes an item, we collapse the rest of the items automatically to have only one delete button visible at a given time. Also you can collapse them by clicking in the main area, but this is used if they don't collapse it by themselves before swiping another.

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/stored/PreselectedStoredPaymentMethodFragment.kt (hunk 3 above)
Comment (ozgur00) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/stored/PreselectedStoredPaymentMethodFragment.kt:172]: Yep checked it with Arjen and decided to do it like this

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodListDialogFragment.kt (hunk 2 above)
Comment (caiofaustino) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodListDialogFragment.kt:187]: Ahh got it, thanks for the clarification!

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt [hunk 4]
```diff
@@ -28,20 +32,36 @@ import com.adyen.checkout.dropin.ui.paymentmethods.PaymentMethodListItem.Compani
 
 @SuppressWarnings("TooManyFunctions")
 class PaymentMethodAdapter(
-    private val paymentMethods: List<PaymentMethodListItem>,
-    private val imageLoader: ImageLoader
+    private val paymentMethods: MutableList<PaymentMethodListItem>,
```
[the remark this answers is not in the document]
Comment (ozgur00) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt:35]: done

[the remark this answers is not in the document]
Comment (ozgur00) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt:35]: done

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt [hunk 5]
```diff
@@ -28,20 +32,36 @@ import com.adyen.checkout.dropin.ui.paymentmethods.PaymentMethodListItem.Compani
...
     fun setPaymentMethodSelectedCallback(onPaymentMethodSelectedCallback: OnPaymentMethodSelectedCallback) {
         this.onPaymentMethodSelectedCallback = onPaymentMethodSelectedCallback
     }
 
+    fun setStoredPaymentRemovedCallback(onStoredPaymentRemovedCallback: OnStoredPaymentRemovedCallback) {
+        this.onStoredPaymentRemovedCallback = onStoredPaymentRemovedCallback
+    }
+
+    fun removePaymentMethodWithId(id: String) {
+        val positionToRemove = paymentMethods.toMutableList().indexOfFirst { it is StoredPaymentMethodModel && it.id == id }
```
[the remark this answers is not in the document]
Comment (ozgur00) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt]: done

File: ui-core/src/main/java/com/adyen/checkout/components/ui/view/AdyenSwipeToRevealLayout.kt [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+import androidx.customview.widget.ViewDragHelper
+import com.adyen.checkout.core.exception.CheckoutException
+import kotlin.math.abs
+import kotlin.math.max
+import kotlin.math.min
+
+private const val CHILD_COUNT = 2
+
+@Suppress("TooManyFunctions")
+class AdyenSwipeToRevealLayout @JvmOverloads constructor(
```
[the remark this answers is not in the document]
Comment (ozgur00) [reply ui-core/src/main/java/com/adyen/checkout/components/ui/view/AdyenSwipeToRevealLayout.kt:51]: done

**Review by caiofaustino — APPROVED** [review]

File: drop-in/src/main/res/layout/removable_payment_methods_list_item.xml [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            android:layout_height="match_parent"
+            android:layout_gravity="center"
+            android:scaleType="centerInside"
+            android:src="@drawable/ic_trash_bin" />
+    </FrameLayout>
+
+    <LinearLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:background="@color/white"
```
[the remark this answers is not in the document]
Comment (ozgur00) [reply drop-in/src/main/res/layout/removable_payment_methods_list_item.xml]: done

**Pull request merged**